card/src/main/java/com/adyen/checkout/card/util/CardNumberUtils.kt

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2023 Adyen N.V.
+ *
+ * This file is open source and available under the MIT license. See the LICENSE file for more info.
+ *
+ * Created by josephj on 31/1/2023.
+ */
+
+package com.adyen.checkout.card.util
+
+import kotlin.math.min
+
+internal object CardNumberUtils {
+
+    fun formatCardNumber(unformattedString: String, maskPartsLengths: List<Int>, separator: String): String {
+        val separatorIndexes = List(maskPartsLengths.size) { index -> maskPartsLengths.take(index + 1).sum() }
+        return unformattedString.foldIndexed("") { index, accumulation, char ->
+            if (separatorIndexes.contains(index)) "$accumulation$separator$char" else accumulation + char
+        }
+
+        return splitStringWithMask(
+            string = unformattedString,
+            maskPartsLengths = maskPartsLengths
+        ).joinToString(separator)
+    }
+
+    private fun splitStringWithMask(string: String, maskPartsLengths: List<Int>): List<String> {
```

Comment:
Did you leave this on purpose? I think the first part seems more readable to me. Only we could add some docs with examples in it.